vega_sim/api/data.py

```diff
@@ -256,25 +256,75 @@ def _transfer_from_proto(
 
 
 def positions_by_market(
-    pub_key: str,
-    market_id: str,
-    position_decimals: int,
-    asset_decimals: int,
-    price_decimals: int,
     data_client: vac.VegaTradingDataClientV2,
-) -> List[Position]:
+    pub_key: str,
+    market_id: Optional[str] = None,
+    position_decimals: Optional[int] = None,
+    asset_decimals: Optional[int] = None,
+    price_decimals: Optional[int] = None,
+) -> Union[Dict[str, Position], Position]:
     """Output positions of a party."""
-    return [
-        _position_from_proto(
-            pos,
-            position_decimals=position_decimals,
-            asset_decimals=asset_decimals,
-            price_decimals=price_decimals,
+
+    raw_positions = data_raw.positions_by_market(
+        pub_key=pub_key, market_id=market_id, data_client=data_client
+    )
+    if len(raw_positions) == 0:
+        logging.debug(
+            f"No positions to return for pub_key={pub_key}, market_id={market_id}"
         )
-        for pos in data_raw.positions_by_market(
-            pub_key=pub_key, market_id=market_id, data_client=data_client
+        return None
+
+    market_price_decimals_map = (
+        {market_id: price_decimals} if market_id is not None else {}
+    )
+    market_position_decimals_map = (
+        {market_id: position_decimals} if market_id is not None else {}
+    )
+    asset_decimals_map = {market_id: asset_decimals} if market_id is not None else {}
+
+    positions = {}
+    for pos in raw_positions:
+
+        market_info = None
+
+        if pos.market_id not in market_price_decimals_map:
+            if market_info is None:
+                market_info = data_raw.market_info(
+                    market_id=pos.market_id, data_client=data_client
+                )
+            market_price_decimals_map[pos.market_id] = int(market_info.decimal_places)
+
+        if pos.market_id not in market_position_decimals_map:
+            if market_info is None:
+                market_info = data_raw.market_info(
+                    market_id=pos.market_id, data_client=data_client
+                )
+            market_position_decimals_map[pos.market_id] = int(
+                market_info.position_decimal_places
+            )
+
+        if pos.market_id not in asset_decimals_map:
+            if market_info is None:
+                market_info = data_raw.market_info(
+                    market_id=pos.market_id, data_client=data_client
+                )
+            asset_info = data_raw.asset_info(
+                asset_id=market_info.tradable_instrument.instrument.future.settlement_asset,
+                data_client=data_client,
+            )
+            asset_decimals_map[pos.market_id] = int(asset_info.details.decimals)
```

Comment:
Not super happy with this but needed for the moment.

(See this [comment](https://github.com/vegaprotocol/vega-market-sim/issues/285#issuecomment-1373858389) in original issue.)